lib/gnat/services/server.ex

```diff
@@ -0,0 +1,167 @@
+
+        # Can also match on endpoint or group
+        def request(msg, "add", "calculator") do
+          {:reply, "42"}
+        end
+
+        # defining an error handler is optional, the default one will just call Logger.error for you
+        def error(%{gnat: gnat, reply_to: reply_to}, _error) do
+          Gnat.pub(gnat, reply_to, "Something went wrong and I can't handle your request")
+        end
+      end
+  """
+alias Gnat.Services.ServiceResponder
+
+  @doc """
+  Called when a message is received from the broker. The endpoint on which the message arrived
+  is always supplied. If the endpoint is a member of a group, the group name will also be
+  provided.
+
+  Automatically increments the request time and processing time stats for this service.
+  """
+  @callback request(message::Gnat.message(), endpoint :: String.t(), group :: String.t() | nil) :: :ok | {:reply, iodata()} | {:error, term()}
+
+  @doc """
+  Called when an error occured during the `request/1`. Automatically increments the error count
+  and processing time stats for this service.
+
+  If your `request/1` function returned `{:error, term}`, then the `term` you returned will be passed as the second argument.
+  If an exception was raised during your `request/1` function, then the exception will be passed as the second argument.
+  If your `request/1` function returned something other than the supported return types, then its return value will be passed as the second argument.
+  """
+  @callback error(message::Gnat.message(), error::term()) :: :ok | {:reply, iodata()}
+
+  defmacro __using__(_opts) do
+    quote do
+      @behaviour Gnat.Services.Server
+
+      def error(_message, error) do
+        require Logger
+        Logger.error(
+          "Gnat.Server encountered an error while handling a request: #{inspect(error)}",
+          type: :gnat_server_error,
+          error: error
+        )
+      end
+
+      defoverridable error: 2
+    end
+  end
+
+
+  @typedoc """
+  Service configuration is provided as part of the consumer supervisor settings in the `service_definition` field.
+  You can specify _either_ the `subscription_topics` field for a reguar server or the `service_definition` field for a
+  new NATS service.
+
+  * `name` - The name of the service. Needs to conform to the rules for NATS service names
+  * `version` - A required version number (w/out "v" prefix) conforming to semver rules
+  * `queue_group` - An optional queue group for service subscriptions. If left off, "q" will be used.
+  * `description` - An optional description of the service
+  * `metadata` - An optional string->string map of service metadata
+  * `endpoints` - A required list of service endpoints. All services must have at least one endpoint
+  """
+  @type service_configuration :: %{
+    required(:name) =>         binary(),
+    required(:version) =>      binary(),
+    required(:endpoints) =>    [endpoint_configuration()],
+    optional(:queue_group) => binary(),
```

Comment:
@autodidaddict I didn't notice this during the first PR, but I'm not sure what the purpose of this service-level queue group would be? It looks like when it's specific you use it in the `Gnat.sub` call, but when I looked at the ADR document, it seems like queue groups only apply to endpoints, not to services?

https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-32.md

I think specifying a queue group for the `$SRV.>` subscription would be an error since we want all instances of the service to decide whether they should respond to a particular services call?

Reply:
This is a good catch. We shouldn't be queue-subscribing on `$SRV` at all.